File: frontend/src/app/position-management/position.service.ts
import { HttpClient } from '@angular/common/http';
import { Injectable, inject } from '@angular/core';
import { Observable, map } from 'rxjs';

export interface PositionItem {
  id: number;
  positionCode: string;
  positionName: string;
  positionRank: number;
  orgLevelScope: string;
  active: boolean;
  createdAt?: string;
  updatedAt?: string;
}

export interface CreatePositionRequest {
  positionCode: string;
  positionName: string;
  positionRank: number;
  orgLevelScope: string;
}

export interface UpdatePositionRequest {
  positionName?: string;
  positionRank?: number;
  orgLevelScope?: string;
  active?: boolean;
}

const BASE_URL = '/excelpro-service/v1/positions';

@Injectable({ providedIn: 'root' })
export class PositionService {
  private readonly http = inject(HttpClient);

  getAll(): Observable<PositionItem[]> {
    return this.http.get<any>(BASE_URL).pipe(map(r => r.data));
  }

  getAllActive(): Observable<PositionItem[]> {
    return this.http.get<any>(`${BASE_URL}/active`).pipe(map(r => r.data));
  }

  create(req: CreatePositionRequest): Observable<PositionItem> {
    return this.http.post<any>(BASE_URL, req).pipe(map(r => r.data));
  }

  update(id: number, req: UpdatePositionRequest): Observable<PositionItem> {
    return this.http.put<any>(`${BASE_URL}/${id}`, req).pipe(map(r => r.data));
  }

  delete(id: number): Observable<void> {
    return this.http.delete<any>(`${BASE_URL}/${id}`).pipe(map(() => void 0));
  }
}

File: backend/src/main/java/org/example/oracleconnectionpool/service/PositionService.java
package org.example.oracleconnectionpool.service;

import lombok.RequiredArgsConstructor;
import org.example.oracleconnectionpool.entity.Position;
import org.example.oracleconnectionpool.model.request.position.CreatePositionRequest;
import org.example.oracleconnectionpool.model.request.position.UpdatePositionRequest;
import org.example.oracleconnectionpool.model.response.PositionResponse;
import org.example.oracleconnectionpool.repository.PositionRepository;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import java.util.List;

@Service
@RequiredArgsConstructor
public class PositionService {

    private final PositionRepository positionRepository;

    public List<PositionResponse> getAll() {
        return positionRepository.findAll().stream()
                .map(this::toResponse)
                .toList();
    }

    public List<PositionResponse> getAllActive() {
        return positionRepository.findByActiveTrueOrderByPositionRankAsc().stream()
                .map(this::toResponse)
                .toList();
    }

    public PositionResponse getById(Long id) {
        Position position = positionRepository.findById(id)
                .orElseThrow(() -> new RuntimeException("Không tìm thấy chức danh với ID: " + id));
        return toResponse(position);
    }

    @Transactional
    public PositionResponse create(CreatePositionRequest request) {
        if (positionRepository.existsByPositionCode(request.getPositionCode())) {
            throw new RuntimeException("Mã chức danh đã tồn tại: " + request.getPositionCode());
        }
        Position position = Position.builder()
                .positionCode(request.getPositionCode().toUpperCase())
                .positionName(request.getPositionName())
                .positionRank(request.getPositionRank())
                .orgLevelScope(request.getOrgLevelScope())
                .active(true)
                .build();
        return toResponse(positionRepository.save(position));
    }

    @Transactional
    public PositionResponse update(Long id, UpdatePositionRequest request) {
        Position position = positionRepository.findById(id)
                .orElseThrow(() -> new RuntimeException("Không tìm thấy chức danh với ID: " + id));

        if (request.getPositionName() != null) position.setPositionName(request.getPositionName());
        if (request.getPositionRank() != null) position.setPositionRank(request.getPositionRank());
        if (request.getOrgLevelScope() != null) position.setOrgLevelScope(request.getOrgLevelScope());
        if (request.getActive() != null) position.setActive(request.getActive());

        return toResponse(positionRepository.save(position));
    }

    @Transactional
    public void delete(Long id) {
        Position position = positionRepository.findById(id)
                .orElseThrow(() -> new RuntimeException("Không tìm thấy chức danh với ID: " + id));
        position.setActive(false);
        positionRepository.save(position);
    }

    private PositionResponse toResponse(Position p) {
        return PositionResponse.builder()
                .id(p.getId())
                .positionCode(p.getPositionCode())
                .positionName(p.getPositionName())
                .positionRank(p.getPositionRank())
                .orgLevelScope(p.getOrgLevelScope())
                .active(p.getActive())
                .createdAt(p.getCreatedAt())
                .updatedAt(p.getUpdatedAt())
                .build();
    }
}
